### PR TITLE
Feature/benchmarks

### PR DESCRIPTION
**Pull request by tmhuysen**
This implements #120

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -95,10 +95,30 @@ install:
     - mkdir build && cd build
     - cmake .. && make -j3 && make test ARGS=-j3 && sudo make install
 
+    # Install google benchmark
```
Comment (lelemmen) [inline .travis.yml]: Good job on installing Google benchmark on Travis. 

Could you or @guacke make sure this installation is also cached? Do we do this in a separate PR after this one, or here?

File: .travis.yml [hunk 2]
```diff
@@ -95,10 +95,30 @@ install:
...
+    - make -j3 && sudo make install
+
 # Run the build script
 script:
   - cd ${TRAVIS_BUILD_DIR}
   - mkdir build && cd build
   - cmake ..
   - make -j3 && make test ARGS=-j3 && sudo make install
-
+  - cd benchmarks
```
Comment (lelemmen) [inline .travis.yml]: Could you add some white lines to separate the testing from the benchmarking?

File: benchmarks/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# benchmark-level CMakeLists.txt
+
+# For every *.cpp-file (every file corresponding to an executable) ...
+foreach(BENCH_SOURCE ${PROJECT_BENCH_SOURCE_FILES})
+
+
+    link_libraries("-lbenchmark")
+    #add_compile_options("-benchmark_counters_tabular=true")
```
Comment (lelemmen) [inline benchmarks/CMakeLists.txt]: Remove this commented line?

File: cmake/SetCMakeEnvironment.cmake [hunk 4]
```diff
@@ -190,6 +190,20 @@ set(PROJECT_EXE_SOURCE_FILES
     ${PROJECT_EXECUTABLES_FOLDER}/oo_doci_rhf.cpp
     ${PROJECT_EXECUTABLES_FOLDER}/fci_lowdin.cpp)
 
+# Find the benchmarks folder
+set(PROJECT_BENCHMARKS_FOLDER ${CMAKE_SOURCE_DIR}/benchmarks)
+
+# Find the source files for the benchmark
```
Comment (lelemmen) [inline cmake/SetCMakeEnvironment.cmake]:
```suggestion
# Find the source files for the benchmarks
```

File: cmake/ConfigureBenchmarks.cmake [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# In this file, we will provide a function that takes the name of an executable, and includes/links the appropriate libraries to it.
```
Comment (lelemmen) [inline cmake/ConfigureBenchmarks.cmake]:
```suggestion
# In this file, we will provide a function that takes the name of a benchmark executable, and includes/links the appropriate libraries to it
```

File: benchmarks/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+    # ... extract the filename without extension (NAME_WE) as a name for our executable ...
+    get_filename_component(EXE_NAME ${BENCH_SOURCE} NAME_WE)
+
+    # ... add an executable based on the source ...
+    add_executable(${EXE_NAME} ${BENCH_SOURCE})
+
+    # ... configure (include headers and link libraries) the executable ...
+    configure_executable(${EXE_NAME})
```
Comment (lelemmen) [inline benchmarks/CMakeLists.txt]: You're using `configure_executable`: shouldn't this be `configure_benchmark_executable`?

File: benchmarks/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,25 @@
+# benchmark-level CMakeLists.txt
+
+# For every *.cpp-file (every file corresponding to an executable) ...
+foreach(BENCH_SOURCE ${PROJECT_BENCH_SOURCE_FILES})
+
+
+    link_libraries("-lbenchmark")
```
Comment (lelemmen) [inline benchmarks/CMakeLists.txt]: I feel like this line should be in the corresponding `cmake/ConfigureBenchmarks.cmake`-file.

File: benchmarks/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    # ... configure (include headers and link libraries) the executable ...
+    configure_executable(${EXE_NAME})
+endforeach()
+
+
+
+# see https://github.com/google/benchmark for program arguments
+# --benchmark_counters_tabular=true (for nice format)
+# --benchmark_out=<filename> (to save to file)
+#
```
Comment (lelemmen) [inline benchmarks/CMakeLists.txt]:
```suggestion
```
but add newline at EOF

File: benchmarks/DOCI/doci_matrix.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+    state.counters["Orbitals"] = K;
+    state.counters["Electron pairs"] = N;
+    state.counters["Dimension"] = fock_space.get_dimension();
+}
+
+
+static void CustomArguments(benchmark::internal::Benchmark* b) {
+    for (int i = 5; i < 9; ++i){
+          b->Args({16,i});
```
Comment (lelemmen) [inline benchmarks/DOCI/doci_matrix.cpp]: Maybe add what this means: 16 orbitals and 5->8 electrons?

File: benchmarks/DOCI/doci_matvec.cpp [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    state.counters["Orbitals"] = K;
+    state.counters["Electron pairs"] = N;
+    state.counters["Dimension"] = fock_space.get_dimension();
+}
+
+
+static void CustomArguments(benchmark::internal::Benchmark* b) {
+    for (int i = 5; i < 9; ++i){
+          b->Args({28,i});
```
Comment (lelemmen) [inline benchmarks/DOCI/doci_matvec.cpp]: idem

File: benchmarks/DOCI/doci_matvec.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    }
+}
+
+// Perform the benchmarks
+BENCHMARK(matvec)->Unit(benchmark::kMillisecond)->Apply(CustomArguments);
+
+
+BENCHMARK_MAIN();
+// --benchmark_counters_tabular=true --benchmark_out=<docimatvec_benchmark>
```
Comment (lelemmen) [inline benchmarks/DOCI/doci_matvec.cpp]: Do you need these lines of code?

File: benchmarks/FCI/fci_matrix.cpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+    state.counters["Orbitals"] = K;
+    state.counters["Electron pairs"] = N;
+    state.counters["Dimension"] = fock_space.get_dimension();
+}
+
+
+static void CustomArguments(benchmark::internal::Benchmark* b) {
+    for (int i = 2; i < 6; ++i){
+          b->Args({8,i});
```
Comment (lelemmen) [inline benchmarks/FCI/fci_matrix.cpp]: idem

File: benchmarks/FCI/fci_matvec.cpp [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+    state.counters["Orbitals"] = K;
+    state.counters["Electron pairs"] = N;
+    state.counters["Dimension"] = fock_space.get_dimension();
+}
+
+
+static void CustomArguments(benchmark::internal::Benchmark* b) {
+    for (int i = 2; i < 6; ++i){
+        b->Args({10,i});
```
Comment (lelemmen) [inline benchmarks/FCI/fci_matvec.cpp]: idem

File: benchmarks/DOCI/doci_matrix.cpp [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+
+static void constructHamiltonian(benchmark::State& state) {
+    // Prepare parameters
+    size_t K = state.range(0);
+    size_t N = state.range(1);
+    GQCP::FockSpace fock_space (K, N);
+    GQCP::DOCI doci (fock_space);
+
+    // Random ham_par
```
Comment (lelemmen) [inline benchmarks/DOCI/doci_matrix.cpp]: over-documentation

File: benchmarks/DOCI/doci_case.cpp [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    auto ham_par = GQCP::readFCIDUMPFile("../../benchmarks/benchmark_input/co_631g_klaas.FCIDUMP");
+
+    // The species contains 14 electrons and 28 basis functions, this requires a single Fock Space of 28 orbitals and 7 electrons
+    GQCP::FockSpace fock_space (ham_par.get_K(), 7);  // dim = 1184040
+
+    // Create the DOCI module
+    GQCP::DOCI doci (fock_space);
+
+    // Solve the Davidson DOCI eigenvalue problem
+    Eigen::VectorXd initial_g = fock_space.HartreeFockExpansion();
```
Comment (lelemmen) [inline benchmarks/DOCI/doci_case.cpp]:
```suggestion
    Eigen::VectorXd initial_guess = fock_space.HartreeFockExpansion();
```

File: benchmarks/DOCI/doci_matvec.cpp [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    // Prepare parameters
+    size_t K = state.range(0);
+    size_t N = state.range(1);
+    GQCP::FockSpace fock_space (K, N);
+    GQCP::DOCI doci (fock_space);
+
+    // Random ham_par
+    GQCP::HamiltonianParameters ham_par = GQCP::constructRandomHamiltonianParameters(K);
+    Eigen::VectorXd diagonal = doci.calculateDiagonal(ham_par);
+    Eigen::VectorXd random = Eigen::VectorXd::Random(diagonal.rows());
```
Comment (lelemmen) [inline benchmarks/DOCI/doci_matvec.cpp]:
```suggestion
    Eigen::VectorXd x =fock_space.randomExpansion();
```

File: benchmarks/FCI/fci_matvec.cpp [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    // Prepare parameters
+    size_t K = state.range(0);
+    size_t N = state.range(1);
+    GQCP::ProductFockSpace fock_space (K, N, N);
+    GQCP::FCI fci (fock_space);
+
+    // Random ham_par
+    GQCP::HamiltonianParameters ham_par = GQCP::constructRandomHamiltonianParameters(K);
+    Eigen::VectorXd diagonal = fci.calculateDiagonal(ham_par);
+    Eigen::VectorXd random = Eigen::VectorXd::Random(diagonal.rows());
```
Comment (lelemmen) [inline benchmarks/FCI/fci_matvec.cpp]:
```suggestion
    Eigen::VectorXd x =fock_space.randomExpansion();
```

File: benchmarks/Hubbard/hubbard_diagonalization.cpp [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+    state.counters["Orbitals"] = K;
+    state.counters["Electron pairs"] = N;
+    state.counters["Dimension"] = fock_space.get_dimension();
+}
+
+
+static void CustomArguments(benchmark::internal::Benchmark* b) {
+    for (int i = 4; i < 10; ++i){
+          b->Args({i,2});
```
Comment (lelemmen) [inline benchmarks/Hubbard/hubbard_diagonalization.cpp]: idem

File: benchmarks/Hubbard/hubbard_matrix.cpp [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+    state.counters["Orbitals"] = K;
+    state.counters["Electron pairs"] = N;
+    state.counters["Dimension"] = fock_space.get_dimension();
+}
+
+
+static void CustomArguments(benchmark::internal::Benchmark* b) {
+    for (int i = 2; i < 5; ++i){
+          b->Args({8,i});
```
Comment (lelemmen) [inline benchmarks/Hubbard/hubbard_matrix.cpp]: idem

File: benchmarks/Hubbard/hubbard_matvec.cpp [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+    // Prepare parameters
+    size_t K = state.range(0);
+    size_t N = state.range(1);
+    GQCP::ProductFockSpace fock_space (K, N, N);
+    GQCP::Hubbard hubbard (fock_space);
+
+    // Random ham_par
+    GQCP::HamiltonianParameters ham_par = GQCP::constructRandomHamiltonianParameters(K);
+    Eigen::VectorXd diagonal = hubbard.calculateDiagonal(ham_par);
+    Eigen::VectorXd random = Eigen::VectorXd::Random(diagonal.rows());
```
Comment (lelemmen) [inline benchmarks/Hubbard/hubbard_matvec.cpp]:
```suggestion
    Eigen::VectorXd x =fock_space.randomExpansion();
```

File: benchmarks/Hubbard/hubbard_matvec.cpp [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+
+    state.counters["Orbitals"] = K;
+    state.counters["Electron pairs"] = N;
+    state.counters["Dimension"] = fock_space.get_dimension();
+}
+
+
+static void CustomArguments(benchmark::internal::Benchmark* b) {
+    for (int i = 2; i < 7; ++i){
+        b->Args({12,i});
```
Comment (lelemmen) [inline benchmarks/Hubbard/hubbard_matvec.cpp]: idem

File: cmake/ConfigureBenchmarks.cmake [hunk 22]
```diff
@@ -7,4 +7,36 @@ function(configure_benchmark_executable EXECUTABLE_NAME)
     target_include_directories(${EXECUTABLE_NAME} PRIVATE ${PROJECT_INCLUDE_FOLDER})
     target_link_libraries(${EXECUTABLE_NAME} PRIVATE ${LIBRARY_NAME})
 
+    # Link google benchmarks
+    link_libraries("-lbenchmark")
+
+    # Include boost
```
Comment (lelemmen) [inline cmake/ConfigureBenchmarks.cmake]: Couldn't you just `configure_executable(${EXECUTABLE NAME})` instead of copy-pasting the code?

File: cmake/ConfigureBenchmarks.cmake [hunk 23]
```diff
@@ -7,4 +7,36 @@ function(configure_benchmark_executable EXECUTABLE_NAME)
     target_include_directories(${EXECUTABLE_NAME} PRIVATE ${PROJECT_INCLUDE_FOLDER})
     target_link_libraries(${EXECUTABLE_NAME} PRIVATE ${LIBRARY_NAME})
 
+    # Link google benchmarks
+    link_libraries("-lbenchmark")
```
Comment (lelemmen) [inline cmake/ConfigureBenchmarks.cmake]: Shouldn't this be `target_link_libraries(${EXECUTABLE_NAME} "-lbenchmark")`?

Comment (lelemmen) [reply cmake/ConfigureBenchmarks.cmake]: Furthermore, can we do `find_package(google_benchmark)`?

Comment (tmhuysen) [reply cmake/ConfigureBenchmarks.cmake]: yes, "find_package(benchmark)" should we "REQUIRE" it? as you don't actually need it (only for benchmarks)

Comment (lelemmen) [reply cmake/ConfigureBenchmarks.cmake]: No, I don't think you should `REQUIRE` it, I think you specify it as `OPTIONAL`.

**Comment by lelemmen** [comment]
@tmhuysen Could you implement the discussed H-chains and H2-chains inside the main library?

@guacke Could you cache the google benchmark compilation?

File: .travis.yml (hunk 1 above)
Comment (lelemmen) [reply .travis.yml]: We'll merge this PR and make an issue later.

File: src/Molecule.cpp [hunk 24]
```diff
@@ -172,6 +172,104 @@ Molecule::Molecule(const std::string& xyz_filename) :
...
+}
+
+
+/**
+ *  @param n            the number of H atoms
+ *  @param spacing      the internuclear spacing in bohr
+ *
+ *  @return a neutral H-chain with equal internuclear spacing
+ */
+Molecule Molecule::HChain(size_t n, double spacing) {
```
Comment (tmhuysen) [inline src/Molecule.cpp]: Why did you use new function instead of charge = 0 default parameter?

Comment (lelemmen) [reply src/Molecule.cpp]: Yes, that was pretty stupid of me. I've updated this in the next commit.
